nimbus/src/lib.rs

```diff
@@ -85,82 +95,103 @@ impl NimbusClient {
         // This is a bit smelly, but get_global_user_participation() needs a
         // writer so that the implementation of update_enrollments can pass one
         // and see the correct value.
-        let db = self.db()?;
+        let db = self.db()?.lock().unwrap();
         let writer = db.write()?;
-        get_global_user_participation(db, &writer)
+        get_global_user_participation(&db, &writer)
     }
 
     pub fn set_global_user_participation(
         &self,
         user_participating: bool,
     ) -> Result<Vec<EnrollmentChangeEvent>> {
-        let db = self.db()?;
+        let db = self.db()?.lock().unwrap();
         let mut writer = db.write()?;
-        set_global_user_participation(db, &mut writer, user_participating)?;
+        set_global_user_participation(&db, &mut writer, user_participating)?;
 
         let existing_experiments = db
             .get_store(StoreId::Experiments)
             .collect_all::<Experiment>(&writer)?;
         // We pass the existing experiments as "updated experiments"
         // to the evolver.
-        let nimbus_id = self.nimbus_id()?;
+        let nimbus_id = self.read_or_create_nimbus_id(&db, &mut writer)?;
         let evolver = EnrollmentsEvolver::new(
             &nimbus_id,
             &self.available_randomization_units,
             &self.app_context,
         );
-        let events = evolver.evolve_enrollments_in_db(db, &mut writer, &existing_experiments)?;
+        let events = evolver.evolve_enrollments_in_db(&db, &mut writer, &existing_experiments)?;
         writer.commit()?;
+        self.database_cache.update(&db)?;
         Ok(events)
     }
 
     pub fn get_active_experiments(&self) -> Result<Vec<EnrolledExperiment>> {
-        get_enrollments(self.db()?)
+        let db = self.db()?.lock().unwrap();
+        get_enrollments(&db)
     }
 
     pub fn get_all_experiments(&self) -> Result<Vec<Experiment>> {
-        self.db()?.collect_all(StoreId::Experiments)
+        let db = self.db()?.lock().unwrap();
+        db.collect_all(StoreId::Experiments)
     }
 
     pub fn opt_in_with_branch(
         &self,
         experiment_slug: String,
         branch: String,
     ) -> Result<Vec<EnrollmentChangeEvent>> {
-        opt_in_with_branch(self.db()?, &experiment_slug, &branch)
+        let db = self.db()?.lock().unwrap();
+        match opt_in_with_branch(&db, &experiment_slug, &branch) {
+            Ok(result) => {
+                self.database_cache.update(&db)?;
+                Ok(result)
+            }
+            Err(e) => Err(e),
+        }
     }
 
     pub fn opt_out(&self, experiment_slug: String) -> Result<Vec<EnrollmentChangeEvent>> {
-        opt_out(self.db()?, &experiment_slug)
+        let db = self.db()?.lock().unwrap();
+        match opt_out(&db, &experiment_slug) {
```

Comment:
Nit: `and_then` might help reduce control-flow noise a bit here:

```
opt_out(&db, &experiment_slug).and_then(|result| {
    self.database_cache.update(&db)?;
    result
})
```

Dunno if having to explicitly pipe through the value of `result` makes it more obscure on balance though. I think you could also use a variable like:

```
let result = opt_out(&db, &experiment_slug)?;
self.database_cache.update(&db)?;
result
```

Reply:
Yeah, I did bounce around a bit here - the `and_then()` returning the original result almost seems to go against what `and_then()` is designed for, and the latter suggestion means we'll try and update the cache even if we failed, which seems like it could be bad - in particular, it might obscure the original error because the cache update seems likely to also fail.

So I left this alone for now, but I don't feel too strongly about it, so let's see if James does :)

Reply:
> the latter suggestion means we'll try and update the cache even if we failed

I don't think it will, the `?` will early-return with an error.

Reply:
doh, right, yeah - that's better (and I don't know how I missed that originally!)